### ISSUE TITLE
Method signature mapping is not deterministic

### ISSUE THREAD

**Issue by jbonjean** [mmnaseri/spring-data-mock#143]
Hi,

It seems there is a random issue related to the `delete` method:

```
Caused by: java.lang.IllegalStateException: Cannot find property `id` on type `class java.lang.Long`
    at [...]ItemServiceTest.delete(ItemServiceTest.java:93)
```

What I understood after analyzing the code, is that the resolver uses introspection to find the matching method but there could be an ambiguity in the method signature because `isAssignableFrom` is used for parameters comparison.

More explicitely, `delete` methods are inherited from `CrudRepository`, the two interesting signatures are:

```
void delete(ID id);
void delete(T entity);
```

Because of type erasure, we can consider that after compilation, it "transformed" to:

```
void delete(Serializable id);
void delete(Object entity);
```

Which is exactly what is declared in the class `DefaultCrudRepository` that the resolver uses to resolve the implementation.

The problem, as I understand it, has two parts:
- The resolver uses `getMethods` and `getDeclaredMethods` to retrieve the method list, which does not guarantee the order (as stated by the documentation "The elements in the array returned are not sorted and are not in any particular order").
- `Object.class.isAssignableFrom(Serializable.class)` is true, which means the resolver will match on the wrong method.

So basically, with current logic based on `getMethods` and `isAssignableFrom`, the method `delete(Serializable id)` maybe mocked to any of the two implementations from `DefaultCrudRepository`.

I found out that when running test from Eclipse (JDT compiler), it is always working properly, maybe because the compiler is more deterministic,  but it is more random when the tests are run from command line.

As a temporary workaround, I changed the resolver to make a first pass that uses `equals` instead of `isAssignableFrom` for parameter comparison, and fallback to the current behaviour. This is not very efficient but fixes the issue.

**Comment by mmnaseri** [comment]
This is an interesting situation. I will try to replicate and fix the behavior properly. Thanks for the time you put into this :)

**Comment by mmnaseri** [comment]
Should be fixed by pull request #147 done by @blackleg

**Comment by edeandrea** [comment]
This same issue holds true on the save method. I am calling CrudRepository.save(Iterable) passing it a List in my code. It ends up calling CrudRepositorySupport.save(Object) rather than DefaultCrudRepository.save(Iterable).

**Comment by blackleg** [comment]
I try to fix this in the pull https://github.com/mmnaseri/spring-data-mock/pull/150

**Comment by usr42** [comment]
I've opened a smaller PR which fixes the issue: https://github.com/mmnaseri/spring-data-mock/pull/160
